### PR TITLE
Include decoder_attention_mask in T5 model inputs

### PR DESCRIPTION
**Pull request by aashiqmuhamed**
# What does this PR do?

This PR includes decoder_attention_mask as an argument in the prepare_inputs_for_generation function, helping enable the use of custom attention masks in the decoder.

Duplicate of https://github.com/huggingface/transformers/pull/22819

@gante @amyeroberts

### CODE REVIEW

**Review by gante — APPROVED** [review]
(normally the doc changes would go into a different PR, but they look fine)

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this! 

For the conflicts, could you rebase on main? There were some recent changes to the docs which should remove the additional diffs and the clashes.

**Comment by aashiqmuhamed** [comment]
@amyeroberts Could you merge this PR please?
